File: src/red_dot_count.py
import os
import cv2
import numpy as np
import pandas as pd
import argparse

def read_tiff_images(tiff_path):
    images = []
    success, images = cv2.imreadmulti(tiff_path, [], cv2.IMREAD_UNCHANGED)
    if not success:
        print(f'오류: {tiff_path} 파일을 읽을 수 없습니다.')
    return images

def count_red_dots(image):
    hsv_image = cv2.cvtColor(image, cv2.COLOR_BGR2HSV)
    lower_red1 = np.array([0, 100, 100])
    upper_red1 = np.array([10, 255, 255])
    lower_red2 = np.array([160, 100, 100])
    upper_red2 = np.array([180, 255, 255])
    mask1 = cv2.inRange(hsv_image, lower_red1, upper_red1)
    mask2 = cv2.inRange(hsv_image, lower_red2, upper_red2)
    red_mask = mask1 | mask2
    return cv2.countNonZero(red_mask)

def process_tiff_file(tiff_path):
    print(f"\n처리 중인 파일: {tiff_path}")
    
    images = read_tiff_images(tiff_path)
    
    print(f'로드된 이미지 수: {len(images)}')
    total_red_dots = 0
    for idx, img in enumerate(images):
        print(f'이미지 {idx + 1}: 크기 {img.shape}, 데이터 타입 {img.dtype}')
        red_dots = count_red_dots(img)
        total_red_dots += red_dots
        print(f'이미지 {idx + 1}의 붉은 점 개수: {red_dots}')
    
    print(f'{os.path.basename(tiff_path)}의 총 붉은 점 개수: {total_red_dots}')
    return total_red_dots

def process_tiff_files_recursively(root_folder):
    for root, dirs, files in os.walk(root_folder):
        tiff_files = [f for f in files if f.lower().endswith(('.tif', '.tiff'))]
        
        if tiff_files:
            red_dot_counts = {}
            for tiff_file in tiff_files:
                full_path = os.path.join(root, tiff_file)
                red_dot_counts[tiff_file] = process_tiff_file(full_path)
            
            if red_dot_counts:
                df = pd.DataFrame(red_dot_counts.items(), columns=['TIFF File', 'Red Dot Count'])
                
                # 'TIFF File' 열에서 '.tiff' 부분 제거
                df['TIFF File'] = df['TIFF File'].str.replace('.tiff', '', regex=False)

                try:
                    # 언더스코어를 기준으로 값을 분리하고 새로운 열 생성
                    split_result = df['TIFF File'].str.split('_', expand=True)
                    if split_result.shape[1] >= 3:
                        df['Number'] = split_result[0]
                        df['Date'] = split_result[1]
                        df['Time'] = split_result[2]
                    else:
                        print(f"경고: {root}의 일부 파일 이름이 예상 형식과 다릅니다.")
                        df['Number'] = df['TIFF File']
                        df['Date'] = pd.NaT
                        df['Time'] = pd.NaT

                    # Date와 Time을 결합하여 datetime 형식으로 변환
                    df['Datetime'] = pd.to_datetime(df['Date'] + df['Time'], format='%Y%m%d%H%M%S', errors='coerce')

                    # Number 열을 정수형으로 변환
                    df['Number'] = pd.to_numeric(df['Number'], errors='coerce')

                    # 필요한 열만 선택
                    df = df[['Datetime', 'Number', 'Red Dot Count']]

                    # Datetime과 Number로 정렬
                    sorted_df = df.sort_values(by=['Datetime', 'Number'])

                    # 인덱스 재설정
                    sorted_df = sorted_df.reset_index(drop=True)

                    # 시작 시간 계산
                    start_time = sorted_df['Datetime'].min()

                    # incubation hour 열 추가
                    sorted_df['incubation hour'] = (sorted_df['Datetime'] - start_time).dt.total_seconds() / 3600

                    csv_path = os.path.join(root, 'red_dot_counts_sorted.csv')
                    sorted_df.to_csv(csv_path, index=False)
                    print(f"\nCSV 파일이 저장되었습니다: {csv_path}")
                except Exception as e:
                    print(f"오류 발생: {root}에서 데이터 처리 중 문제가 발생했습니다.")
                    print(f"오류 메시지: {str(e)}")
                    print("이 폴더의 처리를 건너뛰고 다음 폴더로 진행합니다.")
                    
def main():
    parser = argparse.ArgumentParser(description='TIFF 파일의 붉은 점 개수를 계산하는 스크립트입니다.')
    parser.add_argument('--input', type=str, required=True, help='TIFF 파일이 있는 루트 폴더의 경로')
    args = parser.parse_args()

    print("현재 작업 디렉토리:", os.getcwd())
    print("대상 루트 폴더 경로:", args.input)

    process_tiff_files_recursively(args.input)

if __name__ == "__main__":
    main()

File: src/red_dot_count.rs
use std::fs;
use std::path::{Path, PathBuf};
use std::error::Error;
use opencv::{core, imgcodecs, imgproc, prelude::*};
use chrono::{NaiveDateTime, Duration};
use polars::prelude::*;
use structopt::StructOpt;

#[derive(Debug, StructOpt)]
#[structopt(name = "tiff_processor", about = "TIFF 파일의 붉은 점 개수를 계산하는 스크립트입니다.")]
struct Opt {
    #[structopt(short, long, parse(from_os_str))]
    input: PathBuf,
}

fn read_tiff_images(tiff_path: &Path) -> Result<Vec<Mat>, Box<dyn Error>> {
    let mut images = Vec::new();
    let mut success = false;
    imgcodecs::imreadmulti(tiff_path.to_str().unwrap(), &mut images, imgcodecs::IMREAD_UNCHANGED)?;
    if images.is_empty() {
        println!("오류: {} 파일을 읽을 수 없습니다.", tiff_path.display());
        success = false;
    } else {
        success = true;
    }
    if !success {
        Err("Failed to read TIFF file".into())
    } else {
        Ok(images)
    }
}

fn count_red_dots(image: &Mat) -> Result<i32, Box<dyn Error>> {
    let mut hsv_image = Mat::default();
    imgproc::cvt_color(&image, &mut hsv_image, imgproc::COLOR_BGR2HSV, 0)?;

    let lower_red1 = core::Scalar::new(0.0, 100.0, 100.0, 0.0);
    let upper_red1 = core::Scalar::new(10.0, 255.0, 255.0, 0.0);
    let lower_red2 = core::Scalar::new(160.0, 100.0, 100.0, 0.0);
    let upper_red2 = core::Scalar::new(180.0, 255.0, 255.0, 0.0);

    let mut mask1 = Mat::default();
    let mut mask2 = Mat::default();
    core::in_range(&hsv_image, &lower_red1, &upper_red1, &mut mask1)?;
    core::in_range(&hsv_image, &lower_red2, &upper_red2, &mut mask2)?;

    let mut red_mask = Mat::default();
    core::bitwise_or(&mask1, &mask2, &mut red_mask, &Mat::default())?;

    Ok(core::count_non_zero(&red_mask)?)
}

fn process_tiff_file(tiff_path: &Path) -> Result<i32, Box<dyn Error>> {
    println!("\n처리 중인 파일: {}", tiff_path.display());

    let images = read_tiff_images(tiff_path)?;

    println!("로드된 이미지 수: {}", images.len());
    let mut total_red_dots = 0;
    for (idx, img) in images.iter().enumerate() {
        println!("이미지 {}: 크기 {:?}, 데이터 타입 {:?}", idx + 1, img.size()?, img.typ());
        let red_dots = count_red_dots(img)?;
        total_red_dots += red_dots;
        println!("이미지 {}의 붉은 점 개수: {}", idx + 1, red_dots);
    }

    println!("{}의 총 붉은 점 개수: {}", tiff_path.file_name().unwrap().to_str().unwrap(), total_red_dots);
    Ok(total_red_dots)
}

fn process_tiff_files_recursively(root_folder: &Path) -> Result<(), Box<dyn Error>> {
    for entry in fs::read_dir(root_folder)? {
        let entry = entry?;
        let path = entry.path();
        if path.is_dir() {
            process_tiff_files_recursively(&path)?;
        } else if let Some(extension) = path.extension() {
            if extension.to_str().map(|s| s.to_lowercase()) == Some("tif".to_string())
                || extension.to_str().map(|s| s.to_lowercase()) == Some("tiff".to_string())
            {
                let mut red_dot_counts = Vec::new();
                let red_dot_count = process_tiff_file(&path)?;
                red_dot_counts.push((path.file_name().unwrap().to_str().unwrap().to_string(), red_dot_count));

                if !red_dot_counts.is_empty() {
                    let mut df = DataFrame::new(vec![
                        Series::new("TIFF File", red_dot_counts.iter().map(|(name, _)| name.clone()).collect::<Vec<String>>()),
                        Series::new("Red Dot Count", red_dot_counts.iter().map(|(_, count)| *count).collect::<Vec<i32>>()),
                    ])?;

                    df.apply("TIFF File", |s| s.str()?.replace(".tiff", ""))?;

                    let split_result: Vec<Vec<&str>> = df.column("TIFF File")?
                        .utf8()?
                        .into_iter()
                        .map(|opt_s| opt_s.map(|s| s.split('_').collect()).unwrap_or_default())
                        .collect();

                    if split_result.iter().all(|v| v.len() >= 3) {
                        df.with_column(Series::new("Number", split_result.iter().map(|v| v[0]).collect::<Vec<&str>>()))?;
                        df.with_column(Series::new("Date", split_result.iter().map(|v| v[1]).collect::<Vec<&str>>()))?;
                        df.with_column(Series::new("Time", split_result.iter().map(|v| v[2]).collect::<Vec<&str>>()))?;
                    } else {
                        println!("경고: {}의 일부 파일 이름이 예상 형식과 다릅니다.", path.display());
                        df.with_column(Series::new("Number", df.column("TIFF File")?.clone()))?;
                        df.with_column(Series::new("Date", vec![None; df.height()]))?;
                        df.with_column(Series::new("Time", vec![None; df.height()]))?;
                    }

                    df.with_column(
                        df.column("Date")?
                            .utf8()?
                            .into_iter()
                            .zip(df.column("Time")?.utf8()?.into_iter())
                            .map(|(date, time)| {
                                date.and_then(|d| time.map(|t| format!("{}{}", d, t)))
                                    .and_then(|dt| NaiveDateTime::parse_from_str(&dt, "%Y%m%d%H%M%S").ok())
                            })
                            .collect::<Series>()
                            .rename("Datetime"),
                    )?;

                    df.with_column(
                        df.column("Number")?
                            .cast(&DataType::Int64)
                            .unwrap_or_else(|_| Series::new("Number", vec![None; df.height()])),
                    )?;

                    df = df.select(["Datetime", "Number", "Red Dot Count"])?;

                    df = df.sort(["Datetime", "Number"], vec![false, false])?;

                    let start_time = df.column("Datetime")?.datetime()?.min()?;

                    df.with_column(
                        df.column("Datetime")?
                            .datetime()?
                            .into_iter()
                            .map(|opt_dt| {
                                opt_dt.map(|dt| {
                                    (dt - start_time).num_seconds() as f64 / 3600.0
                                })
                            })
                            .collect::<Series>()
                            .rename("incubation hour"),
                    )?;

                    let csv_path = path.with_file_name("red_dot_counts_sorted.csv");
                    CsvWriter::new(fs::File::create(csv_path)?).finish(&df)?;
                    println!("\nCSV 파일이 저장되었습니다: {}", csv_path.display());
                }
            }
        }
    }
    Ok(())
}

fn main() -> Result<(), Box<dyn Error>> {
    let opt = Opt::from_args();

    println!("현재 작업 디렉토리: {}", std::env::current_dir()?.display());
    println!("대상 루트 폴더 경로: {}", opt.input.display());

    process_tiff_files_recursively(&opt.input)?;

    Ok(())
}